libpromises/evalfunction.c

```diff
@@ -1668,6 +1673,193 @@ static FnCallResult FnCallSplayClass(EvalContext *ctx,
 
 /*********************************************************************/
 
+#ifdef HAVE_LIBCURL
+static size_t cfengine_curl_write_callback(char *ptr, size_t size, size_t nmemb, void *userdata)
+{
+    unsigned int old = BufferSize((Buffer*) userdata);
+    BufferAppend((Buffer*) userdata, ptr, size*nmemb);
+    return BufferSize((Buffer*) userdata) - old;
+}
+#endif
+
+static FnCallResult FnCallCurl(ARG_UNUSED EvalContext *ctx,
+                               ARG_UNUSED const Policy *policy,
+                               ARG_UNUSED const FnCall *fp,
+                               const Rlist *finalargs)
+{
+#ifdef HAVE_LIBCURL
+    if (NULL == CURL_CACHE)
+    {
+        CURL_CACHE = JsonObjectCreate(10);
+    }
+
+    char *url = RlistScalarValue(finalargs);
+    VarRef *ref = ResolveAndQualifyVarName(fp, RlistScalarValue(finalargs->next));
+    if (!ref)
+    {
+        return FnFailure();
+    }
+
+    JsonElement *options = VarRefValueToJson(ctx, fp, ref, NULL, 0);
+    VarRefDestroy(ref);
+
+    if (JsonGetElementType(options) != JSON_ELEMENT_TYPE_CONTAINER ||
+        JsonGetContainerType(options) != JSON_CONTAINER_TYPE_OBJECT)
+    {
+        return FnFailure();
+    }
+
+    Writer *cache_w = StringWriter();
+    WriterWriteF(cache_w, "url = %s; options = ", url);
+    JsonWriteCompact(cache_w, options);
+    JsonElement *old_result = JsonObjectGetAsObject(CURL_CACHE, StringWriterData(cache_w));
+
+    if (NULL != old_result)
+    {
+        Log(LOG_LEVEL_VERBOSE, "%s: found cached request for %s", fp->name, url);
+        WriterClose(cache_w);
+        return (FnCallResult) { FNCALL_SUCCESS, (Rval) { JsonCopy(old_result), RVAL_TYPE_CONTAINER } };
+    }
+
+    if (!CURL_INITIALIZED && curl_global_init(CURL_GLOBAL_DEFAULT) != 0)
+    {
+        Log(LOG_LEVEL_ERR, "%s: libcurl initialization failed, sorry", fp->name);
+
+        WriterClose(cache_w);
+        return FnFailure();
+    }
+
+    //TODO: register curl_global_cleanup() to run on agent exit
+    CURL_INITIALIZED = true;
+
+    CURL *curl = curl_easy_init();
+    if (!curl)
+    {
+        Log(LOG_LEVEL_ERR, "%s: libcurl easy_init failed, sorry", fp->name);
+
+        WriterClose(cache_w);
+        return FnFailure();
+    }
+
+    Buffer *content = BufferNew();
+    Buffer *headers = BufferNew();
+    curl_easy_setopt(curl, CURLOPT_URL, url);
+    // TODO: this uses SIGALRM, is that OK?
```

Comment:
The default action for SIGALRM is termination.
We don't generally catch SIGALRM; in one place we do transiently (and set to SIG_DFL after; this should probably restore whatever signal said was there before, instead); in another place we do catch it but the handler calls alarm(0) as its first action (which may well be non-portable, as some OSen shall block the resulting recursive signal until the handler returns, others shall terminate before running the handler's body).

So: is it OK for whatever's processing this function call to terminate on time-out ?